### PR TITLE
Let CloudProvider return list of NodeAddress, not just one net.IP

### PR DESCRIPTION
**Pull request by justinsb**
This lets cloud providers populate the NodeAddress array

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @ddysher

File: pkg/cloudprovider/cloud.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ type TCPLoadBalancer interface {
 // Instances is an abstract, pluggable interface for sets of instances.
 type Instances interface {
 	// IPAddress returns an IP address of the specified instance.
```
Comment (a-robinson) [inline pkg/cloudprovider/cloud.go]: Need to update the comment as well

File: pkg/cloudprovider/vagrant/vagrant.go [hunk 2]
```diff
@@ -119,14 +119,15 @@ func (v *VagrantCloud) getInstanceByAddress(address string) (*SaltMinion, error)
 	return nil, fmt.Errorf("unable to find instance for address: %s", address)
 }
 
-// IPAddress returns the address of a particular machine instance.
-func (v *VagrantCloud) IPAddress(instance string) (net.IP, error) {
+// NodeAddresses returns the NodeAddress of a particular machine instance.
```
Comment (a-robinson) [inline pkg/cloudprovider/vagrant/vagrant.go]: s/returns the NodeAddress/returns the NodeAddresses

**Comment by a-robinson** [comment]
LGTM other than the comment nits. I don't have the context behind this though, so if @ddysher has anything to say, I'm happy to defer.

File: pkg/cloudprovider/fake/fake.go [hunk 3]
```diff
@@ -115,11 +115,11 @@ func (f *FakeCloud) DeleteTCPLoadBalancer(name, region string) error {
 	return f.Err
 }
 
-// IPAddress is a test-spy implementation of Instances.IPAddress.
-// It adds an entry "ip-address" into the internal method call record.
-func (f *FakeCloud) IPAddress(instance string) (net.IP, error) {
+// NodeAddresses is a test-spy implementation of Instances.NodeAddresses.
+// It adds an entry "node-addresses" into the internal method call record.
+func (f *FakeCloud) NodeAddresses(instance string) ([]api.NodeAddress, error) {
 	f.addCall("ip-address")
```
Comment (ddysher) [inline pkg/cloudprovider/fake/fake.go]: s/ip-address/node-addresses

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 4]
```diff
@@ -270,12 +270,11 @@ func (s *NodeController) PopulateAddresses(nodes *api.NodeList) (*api.NodeList,
...
 			node := &nodes.Items[i]
-			hostIP, err := instances.IPAddress(node.Name)
+			nodeAddresses, err := instances.NodeAddresses(node.Name)
 			if err != nil {
-				glog.Errorf("error getting instance ip address for %s: %v", node.Name, err)
+				glog.Errorf("error getting instance addresses for %s: %v", node.Name, err)
 			} else {
-				address := api.NodeAddress{Type: api.NodeLegacyHostIP, Address: hostIP.String()}
-				api.AddToNodeAddresses(&node.Status.Addresses, address)
+				api.AddToNodeAddresses(&node.Status.Addresses, nodeAddresses...)
```
Comment (ddysher) [inline pkg/cloudprovider/controller/nodecontroller.go]: We should just override it.  `node.status.addresses = nodeAddresses`.  The goal is to have node status re-constructable, if we can't do this, then node addresses shouldn't go into node status at first place.

**Comment by justinsb** [comment]
Thanks for the reviews!  Agree with all suggestions.

I also renamed TestIPAddress -> TestNodeAddresses.  Not sure if that is correct - or whether instead we should preserve the names of tests?

**Comment by a-robinson** [comment]
LGTM, thanks Justin.

**Pull request merged**